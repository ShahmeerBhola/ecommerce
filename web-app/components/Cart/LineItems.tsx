import React, { FC, useContext } from 'react';

import { CartContext, CartItem } from './';
import { Props as CartContentProps } from './Content';
import ContentSection from './ContentSection';

const LineItems: FC<CartContentProps> = ({
    hideDelete,
    hideQuantity,
    showQuantityUnderTitle = false,
    compact = false,
}) => {
    const { lineItems } = useContext(CartContext);

    return lineItems.length > 0 ? (
        <ContentSection text="Your Items" compact={compact}>
            {lineItems.map((lineItem, idx) => (
                <CartItem
                    key={idx}
                    item={lineItem}
                    hideDelete={hideDelete}
                    hideQuantity={hideQuantity}
                    showQuantityUnderTitle={showQuantityUnderTitle}
                    className="p-5 my-2"
                    compact={compact}
                />
            ))}
        </ContentSection>
    ) : null;
};

export default LineItems;
