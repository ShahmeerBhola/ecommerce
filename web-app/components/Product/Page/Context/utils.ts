import isEmpty from 'lodash/isEmpty';

import type { TProductContext } from './types';

export const getSingleAttributes = (_attributes?: StringMap<string[]>): StringMap => {
    let singleAttributes = {};

    for (const [key, value] of Object.entries(_attributes || {})) {
        if (value.length === 1) {
            singleAttributes = {
                ...singleAttributes,
                [key]: value[0],
            };
        }
    }

    return singleAttributes;
};

export const getAvailableAttributes = (
    variations: WordPress.ProductVariation[],
    newAttributes: StringMap,
): StringMap<string[]> => {
    const availableAttribute: StringMap<string[]> = {};

    for (const { attributes } of variations) {
        let isAvailableAttribute = true;

        for (const [key, value] of Object.entries(newAttributes)) {
            isAvailableAttribute = isAvailableAttribute && attributes[key] === value;
        }

        if (isAvailableAttribute) {
            for (const [key, value] of Object.entries(attributes)) {
                const currentAttribute = availableAttribute[key] || [];

                if (currentAttribute.indexOf(value) > -1) {
                    continue;
                }

                availableAttribute[key] = [...currentAttribute, value];
            }
        }
    }

    return availableAttribute;
};

const getAttributes = (
    variations: WordPress.ProductVariation[],
    selectedAttributes: {} | StringMap,
): TProductContext.CheckAttributesType => {
    const availableAttributes = getAvailableAttributes(variations, selectedAttributes);
    const singleAttributes = getSingleAttributes(availableAttributes);

    return {
        currentAttributes: {
            ...selectedAttributes,
            ...singleAttributes,
        },
        availableAttributes,
    };
};

export const checkAttributes = (state: TProductContext.State, payload: any): TProductContext.CheckAttributesType => {
    const { currentAttributes, product } = state;
    const { variations } = product as WordPress.Product;
    const newAttributes = {
        ...currentAttributes,
        ...payload,
    };

    const attributes = getAttributes(variations, newAttributes);

    if (isEmpty(attributes.availableAttributes)) {
        const resetAttributes = {
            ...payload,
        };

        return getAttributes(variations, resetAttributes);
    }

    return attributes;
};

export const isActiveAttribute = (attributes: StringMap, currentAttributes: StringMap): boolean => {
    return Object.entries(attributes).reduce<boolean>((isSelectedAttribute, attribute) => {
        const [key, value] = attribute;
        const matchedAttribute = value === '' || currentAttributes[key] === value;

        return isSelectedAttribute && matchedAttribute;
    }, true);
};
