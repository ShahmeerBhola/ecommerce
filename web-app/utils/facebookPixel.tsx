import React from 'react';
import config from '../config';

const { facebookPixelID } = config;

// https://www.facebook.com/business/help/402791146561655?id=1205376682832142
const event = <P,>(eventName: string, params?: P): void => {
  if (facebookPixelID !== null) {
    if (params !== undefined) {
      window.fbq('track', eventName, params);
    } else {
      window.fbq('track', eventName);
    }
  }
};

export const addLineItemToCart = ({ sku, name, price }: WordPress.ProductBase, type = 'line_item'): void =>
  event<facebook.Pixel.AddToCartParameters>('AddToCart', {
    content_name: name,
    content_type: type,
    content_ids: [sku],
    value: price,
    currency: config.defaultCurrencyCode,
  });

export const addWheelTirePackageToCart = ({ wheel, tire, addOns }: Cart.WheelTirePackage): void => {
  const addWheelTirePackageItemToCart = (product: WordPress.ProductBase): void =>
    addLineItemToCart(product, 'wheel_tire_package');

  addWheelTirePackageItemToCart(wheel);
  addWheelTirePackageItemToCart(tire);
  addOns.map((addOn) => addWheelTirePackageItemToCart(addOn));
};

export const initiateCheckout = (): void => event('InitiateCheckout');
export const completePurchase = (params: facebook.Pixel.PurchaseParameters): void =>
  event<facebook.Pixel.PurchaseParameters>('Purchase', params);

export const FacebookPixelScripts: React.FunctionComponent = () => {
  {
    if (facebookPixelID !== null) {
      return (
        <>
          <script
            dangerouslySetInnerHTML={{
              __html: `
            !function(f,b,e,v,n,t,s)
{if(f.fbq)return;n=f.fbq=function(){n.callMethod ?
              n.callMethod.apply(n, arguments) : n.queue.push(arguments)};
if(!f._fbq)f._fbq=n;n.push=n;n.loaded=!0;n.version='2.0';
n.queue=[];t=b.createElement(e);t.async=!0;
t.src=v;s=b.getElementsByTagName(e)[0];
s.parentNode.insertBefore(t,s)}(window,document,'script',
'https://connect.facebook.net/en_US/fbevents.js');
 fbq('init', '${facebookPixelID}');
fbq('track', 'PageView');
          `,
            }}
          />
          <noscript>
            <img
              height="1"
              width="1"
              src={`https://www.facebook.com/tr?id=${facebookPixelID}&ev=PageView
&noscript=1`}
            />
          </noscript>
        </>
      );
    }
  }
  return null;
};
