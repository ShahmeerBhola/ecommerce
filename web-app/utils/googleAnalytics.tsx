import React from 'react';
import config from '../config';
import { isBrowserSide } from './';

const { googleAnalyticsTrackingID, googleAdsConversionID } = config;

// https://developers.google.com/analytics/devguides/collection/gtagjs/pages
// https://github.com/vercel/next.js/tree/canary/examples/with-google-analytics
export const pageview = (url: string): void => {
  if (googleAnalyticsTrackingID !== null && isBrowserSide()) {
    window.gtag('config', googleAnalyticsTrackingID, {
      page_path: url,
    });
  }
};

// https://developers.google.com/analytics/devguides/collection/gtagjs/events
const event = (action: string, eventParams: any): void => {
  if (googleAnalyticsTrackingID !== null && isBrowserSide()) {
    window.gtag('event', action, eventParams);
  }
};

const conversionEvent = (sendTo: string, eventParams?: StringMap): void =>
  event('conversion', { send_to: sendTo, ...eventParams });

export const sendOrderConfirmationConversionEvent = (orderId: string): void =>
  conversionEvent('AW-479563831/zZZcCPqr3-gBELeg1uQB', { transaction_id: orderId });

export const sendContactFormConversionEvent = (): void => conversionEvent('AW-479563831/6i-DCKyL6-gBELeg1uQB');

export const sendRequestForQuoteEvent = (): void => conversionEvent('AW-479563831/C9PsCJv7_-gBELeg1uQB');

// https://developers.google.com/analytics/devguides/collection/gtagjs/enhanced-ecommerce
export const sendEnhancedEcommerceEvent = (
  eee: GoogleAnalytics.EnhancedEcommerceEvent,
  items: GoogleAnalytics.Product[],
): void => event(eee, { items });

export const sendEnhancedEcommercePurchaseEvent = (data: GoogleAnalytics.PurchaseEvent): void =>
  event('purchase', data);

export const sendEnhancedEcommerceCheckoutOptionEvent = (data: GoogleAnalytics.CheckoutProgressEvent): void =>
  event('set_checkout_option', data);

export const enhancedEcommerceLineItemListName = 'Line Items';
export const enhancedEcommerceWheelTirePackageListName = 'Wheel Tire Package';
export const enhancedEcommerceWheelTirePackageAddOnsListName = 'Wheel Tire Package Add Ons';

export const GoogleAnalyticsAndAdsScripts: React.FunctionComponent = () => {
  {
    if (googleAnalyticsTrackingID !== null && googleAdsConversionID !== null) {
      return (
        <>
          <script async src={`https://www.googletagmanager.com/gtag/js?id=${googleAnalyticsTrackingID}`} />
          <script
            dangerouslySetInnerHTML={{
              __html: `
            window.dataLayer = window.dataLayer || [];
            function gtag(){dataLayer.push(arguments);}
            gtag('js', new Date());
            gtag('config', '${googleAnalyticsTrackingID}', {
              page_path: window.location.pathname,
            });
            gtag('config', '${googleAdsConversionID}')
          `,
            }}
          />
        </>
      );
    }
  }
  return null;
};
