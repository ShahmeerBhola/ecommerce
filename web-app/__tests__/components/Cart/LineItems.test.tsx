import React from 'react';
import renderer from 'react-test-renderer';
import { useCartProvider } from '../../utils';
import { generateMockLineItem, generateMockImage } from '../../utils';
import LineItems from '../../../components/Cart';

describe('LineItems component', () => {
  const lineItems = [
    generateMockLineItem(1234, 25, 'Test Item', 'test-item', 49.99, null),
    generateMockLineItem(1235, 4, 'Test Item 2', 'test-item 2', 599.49, generateMockImage()),
  ];

  test('it renders properly when hideQuantity is true', () => {
    const tree = renderer.create(useCartProvider(<LineItems hideQuantity />, { lineItems })).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('it renders properly when hideQuantity is false', () => {
    const tree = renderer.create(useCartProvider(<LineItems hideQuantity={false} />, { lineItems })).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('it renders properly when hideDelete is true', () => {
    const tree = renderer.create(useCartProvider(<LineItems hideDelete />, { lineItems })).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('it renders properly when hideDelete is false', () => {
    const tree = renderer.create(useCartProvider(<LineItems hideDelete={false} />, { lineItems })).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
