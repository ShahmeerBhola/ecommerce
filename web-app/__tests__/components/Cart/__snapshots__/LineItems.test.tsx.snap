// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LineItems component it renders properly when hideDelete is false 1`] = `
Array [
  <div
    className="flex items-center bg-white my-5 p-5"
  >
    <div
      className="w-1/12 h-20"
    >
      <img
        alt="No Product Image Placeholder"
        className="w-full h-full"
        src="/images/product_image_placeholder.png"
      />
    </div>
    <div
      className="w-6/12 px-3"
    >
      <a
        href="/product/test-item"
        onClick={[Function]}
        onMouseEnter={[Function]}
      >
        <p
          className="font-secondary font-normal text-black text-base font-semibold"
        >
          Test Item
          
        </p>
      </a>
      <p
        className="font-secondary font-normal text-red-100 text-tiny mt-2"
      >
        Size - R20/SX120
      </p>
    </div>
    <div
      className="w-2/12"
    >
      <p
        className="font-secondary font-normal text-gray-100 text-tiny"
      >
        Quantity
      </p>
      <input
        min={1}
        onChange={[Function]}
        type="number"
        value={25}
      />
    </div>
    <div
      className="w-2/12"
    >
      <p
        className="font-secondary font-medium text-red-100 text-base"
      >
        $49.99
      </p>
    </div>
    <div
      className="w-1/12"
    >
      <button
        onClick={[Function]}
      >
        <svg
          fill="none"
          height="14"
          viewBox="0 0 14 14"
          width="14"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M14 2L12 0L7 5L2 0L0 2L5 7L0 12L2 14L7 9L12 14L14 12L9 7L14 2Z"
            fill="#DBDBDB"
          />
        </svg>
      </button>
    </div>
  </div>,
  <div
    className="flex items-center bg-white my-5 p-5"
  >
    <div
      className="w-1/12 h-20"
    >
      <img
        alt=""
        className="w-full h-full"
        src="https://wordpress.standoutspecialties.local/wp-content/uploads/2020/06/CK17-3016-5x500-SF.jpg"
      />
    </div>
    <div
      className="w-6/12 px-3"
    >
      <a
        href="/product/test-item 2"
        onClick={[Function]}
        onMouseEnter={[Function]}
      >
        <p
          className="font-secondary font-normal text-black text-base font-semibold"
        >
          Test Item 2
          
        </p>
      </a>
      <p
        className="font-secondary font-normal text-red-100 text-tiny mt-2"
      >
        Size - R20/SX120
      </p>
    </div>
    <div
      className="w-2/12"
    >
      <p
        className="font-secondary font-normal text-gray-100 text-tiny"
      >
        Quantity
      </p>
      <input
        min={1}
        onChange={[Function]}
        type="number"
        value={4}
      />
    </div>
    <div
      className="w-2/12"
    >
      <p
        className="font-secondary font-medium text-red-100 text-base"
      >
        $599.49
      </p>
    </div>
    <div
      className="w-1/12"
    >
      <button
        onClick={[Function]}
      >
        <svg
          fill="none"
          height="14"
          viewBox="0 0 14 14"
          width="14"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M14 2L12 0L7 5L2 0L0 2L5 7L0 12L2 14L7 9L12 14L14 12L9 7L14 2Z"
            fill="#DBDBDB"
          />
        </svg>
      </button>
    </div>
  </div>,
]
`;

exports[`LineItems component it renders properly when hideDelete is true 1`] = `
Array [
  <div
    className="flex items-center bg-white my-5 p-5"
  >
    <div
      className="w-1/12 h-20"
    >
      <img
        alt="No Product Image Placeholder"
        className="w-full h-full"
        src="/images/product_image_placeholder.png"
      />
    </div>
    <div
      className="w-6/12 px-3"
    >
      <a
        href="/product/test-item"
        onClick={[Function]}
        onMouseEnter={[Function]}
      >
        <p
          className="font-secondary font-normal text-black text-base font-semibold"
        >
          Test Item
          
        </p>
      </a>
      <p
        className="font-secondary font-normal text-red-100 text-tiny mt-2"
      >
        Size - R20/SX120
      </p>
    </div>
    <div
      className="w-2/12"
    >
      <p
        className="font-secondary font-normal text-gray-100 text-tiny"
      >
        Quantity
      </p>
      <input
        min={1}
        onChange={[Function]}
        type="number"
        value={25}
      />
    </div>
    <div
      className="w-2/12"
    >
      <p
        className="font-secondary font-medium text-red-100 text-base"
      >
        $49.99
      </p>
    </div>
  </div>,
  <div
    className="flex items-center bg-white my-5 p-5"
  >
    <div
      className="w-1/12 h-20"
    >
      <img
        alt=""
        className="w-full h-full"
        src="https://wordpress.standoutspecialties.local/wp-content/uploads/2020/06/CK17-3016-5x500-SF.jpg"
      />
    </div>
    <div
      className="w-6/12 px-3"
    >
      <a
        href="/product/test-item 2"
        onClick={[Function]}
        onMouseEnter={[Function]}
      >
        <p
          className="font-secondary font-normal text-black text-base font-semibold"
        >
          Test Item 2
          
        </p>
      </a>
      <p
        className="font-secondary font-normal text-red-100 text-tiny mt-2"
      >
        Size - R20/SX120
      </p>
    </div>
    <div
      className="w-2/12"
    >
      <p
        className="font-secondary font-normal text-gray-100 text-tiny"
      >
        Quantity
      </p>
      <input
        min={1}
        onChange={[Function]}
        type="number"
        value={4}
      />
    </div>
    <div
      className="w-2/12"
    >
      <p
        className="font-secondary font-medium text-red-100 text-base"
      >
        $599.49
      </p>
    </div>
  </div>,
]
`;

exports[`LineItems component it renders properly when hideQuantity is false 1`] = `
Array [
  <div
    className="flex items-center bg-white my-5 p-5"
  >
    <div
      className="w-1/12 h-20"
    >
      <img
        alt="No Product Image Placeholder"
        className="w-full h-full"
        src="/images/product_image_placeholder.png"
      />
    </div>
    <div
      className="w-6/12 px-3"
    >
      <a
        href="/product/test-item"
        onClick={[Function]}
        onMouseEnter={[Function]}
      >
        <p
          className="font-secondary font-normal text-black text-base font-semibold"
        >
          Test Item
          
        </p>
      </a>
      <p
        className="font-secondary font-normal text-red-100 text-tiny mt-2"
      >
        Size - R20/SX120
      </p>
    </div>
    <div
      className="w-2/12"
    >
      <p
        className="font-secondary font-normal text-gray-100 text-tiny"
      >
        Quantity
      </p>
      <input
        min={1}
        onChange={[Function]}
        type="number"
        value={25}
      />
    </div>
    <div
      className="w-2/12"
    >
      <p
        className="font-secondary font-medium text-red-100 text-base"
      >
        $49.99
      </p>
    </div>
    <div
      className="w-1/12"
    >
      <button
        onClick={[Function]}
      >
        <svg
          fill="none"
          height="14"
          viewBox="0 0 14 14"
          width="14"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M14 2L12 0L7 5L2 0L0 2L5 7L0 12L2 14L7 9L12 14L14 12L9 7L14 2Z"
            fill="#DBDBDB"
          />
        </svg>
      </button>
    </div>
  </div>,
  <div
    className="flex items-center bg-white my-5 p-5"
  >
    <div
      className="w-1/12 h-20"
    >
      <img
        alt=""
        className="w-full h-full"
        src="https://wordpress.standoutspecialties.local/wp-content/uploads/2020/06/CK17-3016-5x500-SF.jpg"
      />
    </div>
    <div
      className="w-6/12 px-3"
    >
      <a
        href="/product/test-item 2"
        onClick={[Function]}
        onMouseEnter={[Function]}
      >
        <p
          className="font-secondary font-normal text-black text-base font-semibold"
        >
          Test Item 2
          
        </p>
      </a>
      <p
        className="font-secondary font-normal text-red-100 text-tiny mt-2"
      >
        Size - R20/SX120
      </p>
    </div>
    <div
      className="w-2/12"
    >
      <p
        className="font-secondary font-normal text-gray-100 text-tiny"
      >
        Quantity
      </p>
      <input
        min={1}
        onChange={[Function]}
        type="number"
        value={4}
      />
    </div>
    <div
      className="w-2/12"
    >
      <p
        className="font-secondary font-medium text-red-100 text-base"
      >
        $599.49
      </p>
    </div>
    <div
      className="w-1/12"
    >
      <button
        onClick={[Function]}
      >
        <svg
          fill="none"
          height="14"
          viewBox="0 0 14 14"
          width="14"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M14 2L12 0L7 5L2 0L0 2L5 7L0 12L2 14L7 9L12 14L14 12L9 7L14 2Z"
            fill="#DBDBDB"
          />
        </svg>
      </button>
    </div>
  </div>,
]
`;

exports[`LineItems component it renders properly when hideQuantity is true 1`] = `
Array [
  <div
    className="flex items-center bg-white my-5 p-5"
  >
    <div
      className="w-1/12 h-20"
    >
      <img
        alt="No Product Image Placeholder"
        className="w-full h-full"
        src="/images/product_image_placeholder.png"
      />
    </div>
    <div
      className="w-6/12 px-3"
    >
      <a
        href="/product/test-item"
        onClick={[Function]}
        onMouseEnter={[Function]}
      >
        <p
          className="font-secondary font-normal text-black text-base font-semibold"
        >
          Test Item
          
        </p>
      </a>
      <p
        className="font-secondary font-normal text-red-100 text-tiny mt-2"
      >
        Size - R20/SX120
      </p>
    </div>
    <div
      className="w-2/12"
    >
      <p
        className="font-secondary font-medium text-red-100 text-base"
      >
        $49.99
      </p>
    </div>
    <div
      className="w-1/12"
    >
      <button
        onClick={[Function]}
      >
        <svg
          fill="none"
          height="14"
          viewBox="0 0 14 14"
          width="14"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M14 2L12 0L7 5L2 0L0 2L5 7L0 12L2 14L7 9L12 14L14 12L9 7L14 2Z"
            fill="#DBDBDB"
          />
        </svg>
      </button>
    </div>
  </div>,
  <div
    className="flex items-center bg-white my-5 p-5"
  >
    <div
      className="w-1/12 h-20"
    >
      <img
        alt=""
        className="w-full h-full"
        src="https://wordpress.standoutspecialties.local/wp-content/uploads/2020/06/CK17-3016-5x500-SF.jpg"
      />
    </div>
    <div
      className="w-6/12 px-3"
    >
      <a
        href="/product/test-item 2"
        onClick={[Function]}
        onMouseEnter={[Function]}
      >
        <p
          className="font-secondary font-normal text-black text-base font-semibold"
        >
          Test Item 2
          
        </p>
      </a>
      <p
        className="font-secondary font-normal text-red-100 text-tiny mt-2"
      >
        Size - R20/SX120
      </p>
    </div>
    <div
      className="w-2/12"
    >
      <p
        className="font-secondary font-medium text-red-100 text-base"
      >
        $599.49
      </p>
    </div>
    <div
      className="w-1/12"
    >
      <button
        onClick={[Function]}
      >
        <svg
          fill="none"
          height="14"
          viewBox="0 0 14 14"
          width="14"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M14 2L12 0L7 5L2 0L0 2L5 7L0 12L2 14L7 9L12 14L14 12L9 7L14 2Z"
            fill="#DBDBDB"
          />
        </svg>
      </button>
    </div>
  </div>,
]
`;
